auth/oci.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"os/user"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-ini/ini"
+)
+
+const OCIConfigPath = ".oraclebmc/mantle"
+
+// OCIProfile represents a parsed OCI profile.
+type OCIProfile struct {
+	TenancyID   string `ini:"tenancy"`
+	UserID      string `ini:"user"`
+	Fingerprint string `ini:"fingerprint"`
+	KeyFile     string `ini:"key_file"`
+	Region      string `ini:"region"`
+}
+
+// ReadOCIConfig decodes an OCI config file, for more information:
+// https://docs.us-phoenix-1.oraclecloud.com/Content/API/Concepts/sdkconfig.htm
+//
+// If path is empty, $HOME/.oraclebmc/mantle is read.
+func ReadOCIConfig(path string) (map[string]OCIProfile, error) {
+	if path == "" {
+		user, err := user.Current()
+		if err != nil {
+			return nil, err
+		}
+		path = filepath.Join(user.HomeDir, OCIConfigPath)
+	}
+
+	profiles := make(map[string]OCIProfile)
+	cfg, err := ini.InsensitiveLoad(path)
+	if err != nil {
+		return nil, fmt.Errorf("Loading OCI config: %v", err)
+	}
+
+	for _, section := range cfg.Sections() {
+		p := OCIProfile{}
+		err = section.MapTo(&p)
+		if err != nil {
+			return nil, err
+		}
+
+		profiles[strings.ToLower(section.Name())] = p
+	}
+	if len(profiles) == 0 {
+		return nil, fmt.Errorf("OCI config %q contains no profiles", path)
+	}
+
+	return profiles, nil
+}
+
+func (profile *OCIProfile) GetPrivateKey() (*rsa.PrivateKey, error) {
+	keyData, err := ioutil.ReadFile(profile.KeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("reading RSA key: %v", err)
+	}
+
+	block, _ := pem.Decode(keyData)
+	if block == nil {
+		return nil, fmt.Errorf("parsing PEM block")
+	}
+
+	if priv, err := x509.ParsePKCS1PrivateKey(block.Bytes); err != nil {
+		return nil, fmt.Errorf("parsing RSA key: %v", err)
+	} else {
+		return priv, nil
+	}
+}
```

Comment:
Sadly, we should probably check that the key matches the fingerprint, since otherwise there will be confusion later.

Reply:
Working on adding this back in, the tests in `platform/api/oci/client_test.go` will have to be updated as a result because they're using the example data provided by Oracle in the docs and the provided RSA key does not match the provided fingerprint they use in the KeyID.